src/crimson/os/seastore/seastore.cc

```diff
@@ -400,23 +400,127 @@ seastar::future<store_statfs_t> SeaStore::stat() const
+
+    uint64_t end_hash = reverse_hash  + (1ull << (32 - bits));
+    if (end_hash > MAX_HASH) {
+      // make sure end hobj is even greater than the maximum possible hobj
+      end->hobj.set_bitwise_key_u32(MAX_HASH);
+      temp_end->hobj.set_bitwise_key_u32(MAX_HASH);
+      end->hobj.nspace = MAX_NSPACE;
+    } else {
+      end->hobj.set_bitwise_key_u32(end_hash);
+      temp_end->hobj.set_bitwise_key_u32(end_hash);
+    }
+  } else {
+    start->shard_id = shard_id_t::NO_SHARD;
+    start->hobj.pool = -1ull;
+
+    *end = *start;
+    start->hobj.set_bitwise_key_u32(0);
+    end->hobj.set_bitwise_key_u32(MAX_HASH);
+    end->hobj.nspace = MAX_NSPACE;
+    // no separate temp section
+    *temp_start = *end;
+    *temp_end = *end;
+  }
+
+  start->generation = 0;
+  end->generation = 0;
+  temp_start->generation = 0;
+  temp_end->generation = 0;
+}
+
+TransactionManager::read_extent_iertr::future<std::optional<unsigned>>
+SeaStore::get_coll_bits(CollectionRef ch, Transaction &t) const
+{
+  return transaction_manager->read_collection_root(t)
+    .si_then([this, ch, &t](auto coll_root) {
+      return collection_manager->list(coll_root, t);
+    }).si_then([ch](auto colls) {
+      auto it = std::find_if(colls.begin(), colls.end(),
+        [ch](const std::pair<coll_t, coll_info_t>& element) {
+          return element.first == ch->get_cid();
+      });
+      if (it != colls.end()) {
+        return TransactionManager::read_extent_iertr::make_ready_future<
+          std::optional<unsigned>>(it->second.split_bits);
+      } else {
+        return TransactionManager::read_extent_iertr::make_ready_future<
+	  std::optional<unsigned>>(std::nullopt);
+      }
+    });
+}
+
+OnodeManager::RangeTuple
+SeaStore::get_objs_range(CollectionRef ch, unsigned bits) const
+{
+  ghobject_t coll_range_temp_start, coll_range_temp_end;
+  ghobject_t coll_range_start, coll_range_end;
+  get_coll_range(ch->get_cid(),
+                 bits,
+                 &coll_range_temp_start,
+                 &coll_range_temp_end,
+                 &coll_range_start,
+                 &coll_range_end);
+  return std::make_tuple(coll_range_temp_start,
+                         coll_range_temp_end,
+                         coll_range_start,
+                         coll_range_end);
+}
+
 seastar::future<std::tuple<std::vector<ghobject_t>, ghobject_t>>
 SeaStore::list_objects(CollectionRef ch,
-                        const ghobject_t& start,
-                        const ghobject_t& end,
-                        uint64_t limit) const
+                       const ghobject_t start,
+                       const ghobject_t end,
+                       uint64_t limit) const
 {
   using RetType = typename OnodeManager::list_onodes_bare_ret;
+  if (start.is_max() || start.hobj.is_max()) {
+    return seastar::make_ready_future<
+      std::tuple<std::vector<ghobject_t>, ghobject_t>>(
+      std::make_tuple(std::vector<ghobject_t>(), ghobject_t::get_max()));
+  }
```

Comment:
This check doesn't seem necessary, if start.is_max(), start will be to the right of the collection boundaries anyway.

Reply:
I keep the boundary check at beginning to return directly and not go below to do transaction and repeat.

Reply:
do you mean if start is max, still need go to get_obj_ranges and get_ranges to calculate ranges and get a empty ranges?  why not return empty vector directly here?

Reply:
Duplicating logic unnecessarily creates extra opportunities for bugs.  Future changes to this method that affect the behavior at the boundaries would need to fix this check in addition to get_obj_ranges -- that might not be obvious at the time.

It complicates testing as well.  If this case is handled naturally as part of a larger class of inputs, it's less important to test it specifically.  If we leave this short-circuit here, though, I'd want you to add a test case for it specifically.  I'd also want you to  add a test for a value of start that is to the right of the max object ghobject but not is_max() since passing ghobject_t::is_max() will hit this short circuit rather than testing the more general pathway.

It's not worth much as an optimization.  Users of list_objects() would naturally stop iterating once start.is_max(), so this short circuit should be essentially dead code.  (If there's some hot path user that really does pass in a max start value, it might be worthwhile, though).

Reply:
I see, I will move boundary check to get_ranges and return empty ranges for start is max.

Reply:
I am a little confused here seems I only need remove the boundary check is ok, do you want me add the check in get_ranges?

Reply:
if onode_manager->list_onodes get max as start will return empty vector result, so needn't check boundary here.

Reply:
Right, I think all you need to do here is to remove this extra check -- the rest of the implementation should just work.